File: BFCodeGen.hs
module BFCodeGen where
import Data.List
import Data.Ord
import Data.Maybe
import Data.Char

x ^-^ y = x `mod` y

compareCase :: (Integral a) => (a,a,a,a) -> a
compareCase (a,b,s,k) =
    let extras = if a+b > 0 then 6 else 0
    in a + b + (abs s) + (abs k) + extras

calc :: (Integral a) => a -> [(a,a,a,a)] -> Maybe (a,a,a,a)
calc c allCases =
    if c == 0
        then Just (0,0,0,0)
    else
        let workCases = filter (\(a,b,s,t) -> t * b `mod` 256 == c) allCases
            useCases = map (\(a,b,s,t) -> (a,b,s,0)) workCases
        in case useCases of
            [] -> Nothing
            _ -> Just (minimumBy (comparing compareCase) useCases)

slice :: Int -> Int -> [a] -> [a]
slice from to xs = take (to - from) (drop from xs)

compareMaybeCase :: (Integral a) => Maybe (a,a,a,a) -> Maybe (a,a,a,a) -> Ordering
compareMaybeCase Nothing Nothing = EQ
compareMaybeCase (Just vala) Nothing = GT
compareMaybeCase Nothing (Just valb) = LT
compareMaybeCase (Just a) (Just b) = compare (compareCase a) (compareCase b)

get_nei_range :: Int -> Int -> Int -> (Int, Int)
get_nei_range index span lencs =
    (l,r)
    where
        l = max 0 (index - span + 1)
        r = min lencs (index + span)

best_neis_pos :: (Integral a) => Int -> [Maybe (a,a,a,a)] -> [(Maybe (a,a,a,a), a)]
best_neis_pos index cs =
    zip subCs pos
    where
        span = 10
        (l,r) = get_nei_range index span $ length cs
        lshift = l - index
        rshift = r - index
        subCs = slice l r cs
        pos = slice lshift rshift $ (reverse $ take (max 0 (-lshift)) [-1,-2..]) ++ [0..]

best_neis :: (Integral a) => Int -> [Maybe (a,a,a,a)] -> Maybe (a,a,a,a)
best_neis index cs =
    minimumBy compareMaybeCase workNeis
    where
        neis = map (\(val, shift) -> case val of
            Just (a,b,s,k) -> Just (a,b,s, k - shift)
            Nothing -> Nothing) $ best_neis_pos index cs
        workNeis = filter(\nei -> case nei of
            Just _ -> True
            Nothing -> False) neis

lastN :: Int -> [a] -> [a]
lastN n xs = drop (max ((length xs) - n) 0) xs

best_left :: (Integral a) => [Maybe (a,a,a,a)] -> [Maybe (a,a,a,a)]
best_left_fn :: (Integral a) => [Maybe (a,a,a,a)] -> [Maybe (a,a,a,a)] -> [Maybe (a,a,a,a)]
best_left_fn last10 value =
    lastN 10 (last10 ++ [curr_best])
    where   values    = last10 ++ value
            curr_best = best_neis (length last10) values
best_left cs = map last (scanl1 best_left_fn $ map (\c -> [c]) cs)
best_left_test cs = scanl1 best_left_fn $ map (\c -> [c]) cs

best_right :: (Integral a) => [Maybe (a,a,a,a)] -> [Maybe (a,a,a,a)]
best_right_fn :: (Integral a) => [Maybe (a,a,a,a)] -> [Maybe (a,a,a,a)] -> [Maybe (a,a,a,a)]
best_right_fn value next10 =
    take 10 ([curr_best] ++ next10)
    where
        values    = value ++ next10
        curr_best = best_neis 0 values
best_right cs = map head (scanr1 best_right_fn $ map (\c -> [c]) cs)
best_right_test cs = scanr1 best_right_fn $ map (\c -> [c]) cs

get_code:: (Integral a) => a -> [Char]
get_code n = if n >= 0 then (take (fromIntegral n) (repeat '+')) else (take (-(fromIntegral n)) (repeat '-'))

gen_code :: (Integral a) => Char -> [(a,a,a,a)] -> [Char]
gen_code c best_c =
    let index = ord c
        (a,b,s,k) = best_c!!index
        s_code = get_code s
        loop_code = if a+b >0 then ("[" ++ (get_code (-a)) ++ ">" ++ (get_code b) ++ "<]>") else ""
        k_code = get_code k
    in s_code ++ loop_code ++ k_code ++ "."

File: BF_code_gen.py
best = [(-1,0,0,0)] * 256
init = False

def cost(a,b,s,k):
    extras = 6 if a+b > 0 else 0
    return a+b+abs(s) + abs(k) + extras

def calc(c):
    if c == 0:
        return (0,0,0,0)
    found = False
    for a in range(2,51):
        for s in range(-50, 51):
            if s == 0:
                continue
            for t in range(0,256):
                if (-t*a+s) % 256 == 0:
                    for b in range(0,51):
                        if t * b % 256 != c:
                            continue
                        found = True
                        if best[c][0] == -1 or cost(*best[c]) > cost(a,b,s,0):
                            best[c] = (a,b,s,0)
                    break
    if not found:
        print(f"c: {c}, Not found!")

def run():
    best[0] = (0,0,0,0)
    for c in range(103, 103+256):
        calc(c%256)
        a,b,s,k = best[c%256]

    mx = 0
    mn = float('inf')
    best_left = [(-1,0,0,0)] * 256
    best_right = [(-1,0,0,0)] * 256
    for c in range(0, 256):
        span = 10
        best_left[c] = best[c]
        for i in range(-span,0):
            if 0 <= c + i < 256:
                a,b,s,k = best_left[c]
                ai,bi,si,ki = best_left[c+i]
                if ai + bi + abs(si) +abs(ki-i) < a + b + abs(s) + abs(k) or (a == -1 and ai != -1):
                    best_left[c] = (ai,bi,si,ki-i) 
    
    for c in range(255,-1,-1):
        span = 10
        best_right[c] = best[c]
        for i in range(1,span):
            if 0 <= c + i < 256:
                a,b,s,k = best_right[c]
                ai,bi,si,ki = best_right[c+i]
                if ai + bi + abs(si) +abs(ki-i) < a + b + abs(s) + abs(k) or (a == -1 and ai != -1):
                    best_right[c] = (ai,bi,si,ki-i)

    for c in range(0,256):
        best[c] = best_left[c]
        al,bl,sl,kl = best_left[c]
        ar,br,sr,kr = best_right[c]
        if al + bl + abs(sl) + abs(kl) < ar + br + abs(sr) + abs(kr):
            best[c] = best_left[c]
        else:
            best[c] = best_right[c]
        a,b,s,k = best[c]

def gen(c):
    global init, best
    if not init:
        run()
        init = True
    a,b,s,k = best[c]
    if s >= 0:
        print('+'*(s), end='')
    else:
        print('-'*(-s), end='')
    if a + b > 0:
        print('[', end='')
        if a:
            print('-'*a, end='')
        if b:
            print('>' + '+'*b + '<', end='')
        print(']>', end= '')
    if k >= 0:
        print('+'*(k), end='')
    else:
        print('-'*(-k), end='')
    print('.')